### ISSUE TITLE
Bug: False positive in no-unmodified-loop-condition with LogicalExpressions (&&, ||)

### ISSUE THREAD

**Issue by Kuldeep2822k** [eslint/eslint#20844]
### Environment

Node version: v24.15.0
npm version: v11.12.1
Local ESLint version: v10.2.1
Global ESLint version: N/A
Operating System: Windows 11

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
// eslint.config.js
module.exports = [
    {
        rules: {
            "no-unmodified-loop-condition": "error"
        }
    }
];
```
</details>

```js
function testFunction(filter) {
    let obj = { value: 1 };
    
    do {
        obj = Object.getPrototypeOf(obj);
    } while (obj && !filter);
    
    return obj;
}
```

### What did you expect to happen?

No error should be reported. 

The variable `obj` is modified inside the loop, guaranteeing that the loop is not infinite. The `!filter` variable is safely used as a constant guard. 

There is no logical difference between `obj !== filter` (which correctly reports no error because the rule parses BinaryExpressions correctly) and `obj && !filter` in terms of loop termination when one operand is modified.

### What actually happened?

The rule throws a false positive:
`error  'filter' is not modified in this loop  no-unmodified-loop-condition`

**Suspected Cause:**
It looks like `LogicalExpression` might be missing from the `GROUP_PATTERN` regex in `lib/rules/no-unmodified-loop-condition.js` (which currently handles `BinaryExpression` and `ConditionalExpression`). 

If `LogicalExpression` isn't grouped, it seems the rule evaluates the `!filter` branch in complete isolation from the modified `obj` branch, which leads to the false positive.

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24gdGVzdEZ1bmN0aW9uKGZpbHRlcikge1xuICAgIGxldCBvYmogPSB7IHZhbHVlOiAxIH07XG4gICAgXG4gICAgZG8ge1xuICAgICAgICBvYmogPSBPYmplY3QuZ2V0UHJvdG90eXBlT2Yob2JqKTtcbiAgICB9IHdoaWxlIChvYmogJiYgIWZpbHRlcik7XG4gICAgXG4gICAgcmV0dXJuIG9iajtcbn1cbiIsIm9wdGlvbnMiOnsicnVsZXMiOnsibm8tdW5tb2RpZmllZC1sb29wLWNvbmRpdGlvbiI6WyJlcnJvciJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX19fX0=

### Participation

- [x] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

While looking into this locally, I noticed that the `GROUP_PATTERN` regex inside `lib/rules/no-unmodified-loop-condition.js` currently looks like this:
`/^(?:BinaryExpression|ConditionalExpression)$/u`

I suspect that adding `LogicalExpression` to this group pattern might resolve the issue. If the team agrees this is a bug and that this is the correct approach, I would be very happy to submit a Pull Request to implement the fix and add the corresponding unit tests!

**Comment by DMartens** [comment]
I think the rule is working as intended.
The intent of the rule is not to detect infinite loops but probable logical errors.
In this case including the variable `filter` is unnecessary as it is either constantly `true` or `false`:
- if it is `true` it can be omitted as it does not change the outcome of the test condition
- if it is `false` the loop only runs once, it is a `do-while` loop as the test condition evaluates to `false` regardless of the rest of the condition

**Comment by Kuldeep2822k** [comment]
Thank you for the reply and the explanation!

I see your point about why a constant condition could be seen as a logical error if it were a hardcoded literal. However, in the example, `filter` is a **function parameter**. 

```javascript
function testFunction(filter) {  // <-- filter is a parameter
    let obj = { value: 1 };
    do {
        obj = Object.getPrototypeOf(obj);
    } while (obj && !filter); 
    return obj;
}
```
While `filter` remains unmodified *during the loop execution*, it is not an unnecessary condition. If the function is called with `testFunction(true)`, the loop runs once. If called with `testFunction(false)`, it traverses the prototype chain. Using an unmodified parameter alongside a modified variable is a very standard pattern for early-exits or feature flags.

Additionally, my main concern is the **inconsistency** in how the rule applies this logic. Currently, the rule completely allows unmodified parameters when used in a `BinaryExpression`, because `BinaryExpression` is grouped in the `GROUP_PATTERN` regex. 

For example, this code throws **no errors** because of the `!==` operator, even though `filter` is just as unmodified here as it is with `&&`:
```javascript
function testFunction(filter) {
    let obj = { value: 1 };
    
    // NO ERROR reported for 'filter'. 
    do {
        obj = Object.getPrototypeOf(obj);
    } while (obj !== filter); 
    
    return obj;
}
```

If the rule allows `while (obj !== filter)` because `obj` is modified (which it currently does), it seems logically consistent that it should also allow `while (obj && !filter)`. Adding `LogicalExpression` to the group pattern resolves this inconsistency so that `&&` and `!==` are treated by the same rules.

Does that help clarify the use case?

**Comment by DMartens** [comment]
It does not matter whether `filter` is a parameter as in this case it only matters for the rule whether it is truthy or falsy.
For example the following code is equivalent and in my opinion easier to understand, as the code flow is more explicit:
```javascript
function testFunction(filter) {
    let obj = { value: 1 };
    if (!filter) {
        return Object.getPrototypeOf(obj);
    }
    
    do {
        obj = Object.getPrototypeOf(obj);
    } while (obj);
    
    return obj;
}
```

**Comment by Kuldeep2822k** [comment]
You're right that it doesn't actually demonstrate what I said it did:

```jsx
async function processStream(stream, shouldAbort) {
    let chunk = stream.read();
    while (chunk && !shouldAbort) {
        await process(chunk);
        chunk = stream.read();
    }
}
```

`shouldAbort` is a parameter that's never reassigned, so the rule flagging it is correct. My bad.

That said, I do still think there's something off here, just not where I originally pointed. Let me try a cleaner pair of examples.

**Case A — `&&`:**

```jsx
function nextOrNull() {
    return Math.random() > 0.5 ? { data: 1 } : null;
}

let chunk = nextOrNull();
let done = false;

while (chunk && !done) {
    chunk = nextOrNull(); // can become null, loop can terminate
}
```

ESLint:

```
8:18  error  'done' is not modified in this loop  no-unmodified-loop-condition
```

**Case B — `?:` with the same termination logic:**

```jsx
function nextOrNull() {
    return Math.random() > 0.5 ? { data: 1 } : null;
}

let chunk = nextOrNull();
let done = false;

while (chunk ? !done : false) {
    chunk = nextOrNull();
}
```

ESLint: no errors.

Both loops can exit the same way (when `chunk` becomes null), and `done` is unmodified in both. Only the operator changes, but the rule's verdict flips.

Looking at the rule source, this falls out of `GROUP_PATTERN` on line 29:

```jsx
const GROUP_PATTERN = /^(?:BinaryExpression|ConditionalExpression)$/u;

/**
 * @property {ASTNode} group - BinaryExpression or ConditionalExpression nodes
 *      that the reference is belonging to.
 */
```

When a reference's parent matches `GROUP_PATTERN`, the rule checks whether *any* reference in the group is modified, and if so, the whole group is treated as safe. `LogicalExpression` isn't in there, so `!done` in `chunk && !done` gets checked on its own and trips the warning. The `?:` version is grouped with `chunk`, so it passes.

One thing I want to address up front, since I'd ask the same thing: why extend `GROUP_PATTERN` rather than shrink it? Removing `BinaryExpression`/`ConditionalExpression` would regress the canonical `while (i < n)` case (where `n` is a cached `array.length` or similar bound). The grouping was added to avoid exactly that. Adding `LogicalExpression` keeps that working and just restores symmetry across the three operator forms.

To be clear, I'm not arguing the rule should stop flagging unmodified loop variables in general — it should. The narrower claim is just that the same identifier in logically equivalent expressions shouldn't get different verdicts based on which operator combines it with the modified one.

Would you be open to looking at whether `LogicalExpression` belongs in `GROUP_PATTERN` next to the other two?

**Comment by DMartens** [comment]
The mentioned cases work as expected.
Case A:
```javascript
function nextOrNull() {
    return Math.random() > 0.5 ? { data: 1 } : null;
}

let chunk = nextOrNull();
let done = false;

while (chunk) { // "done" can simply be removed as it is not modified
    chunk = nextOrNull(); // can become null, loop can terminate
}
```

Case B:
```javascript
function nextOrNull() {
    return Math.random() > 0.5 ? { data: 1 } : null;
}

let chunk = nextOrNull();
let done = false;

while (chunk ? !done : false) { // chunk is the test condition which changes (the rule does not check for "done" --> false negative
    chunk = nextOrNull();
}
```

Yes, the conditions for case A and B are equivalent but I argue that the behavior for the logical expression is still correct as I would classify the ternary in class B as a false negative.

**Comment by Kuldeep2822k** [comment]
- Thanks again for clarifying this issue.
- I won’t pursue the LogicalExpression change further here.
- Before opening anything new: would you prefer I open a separate issue for the ConditionalExpression false-negative case, or leave it as current behavior?

**Comment by DMartens** [comment]
There is no need for a separate issue; feel free to open a PR for fixing the `ConditionalExpression` false negative.

**Comment by ivangray271-lgtm** [comment]
Thank you for the guidance.

**Comment by sethamus** [comment]
I would like to work on this.

**Comment by DMartens** [comment]
Feel free to open a PR for this @sethamus. This should be implemented as an additional option ([context](https://github.com/eslint/eslint/pull/20864#discussion_r3302610685)).
